### ISSUE TITLE
Importing SynapseML doesn't work

### ISSUE THREAD

**Issue by vishalovercome** [microsoft/pylance-release#4648]
I'm working on a pyspark project and trying to leverage Microsoft's SynapseML library. There are two ways of installing synapseml:

1. Dowloading as a JAR when running Spark submit - This is how it works in production but after downloading it locally, I'm unable to import the python libraries dues to which none of the IDE features work
2. Installing it via pip - This won't work in production (spark-submit, Jupyter, etc.) but at least lets us leverage IDE features. 

Is there a way to get all IDE features without installing it via pip?

**Comment by debonte** [comment]
There's documentation on how we resolve imports [here](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md). Try setting [`python.analysis.extraPaths`](https://github.com/microsoft/pyright/blob/main/docs/settings.md) in your VS Code settings or [`extraPaths`](https://github.com/microsoft/pyright/blob/main/docs/configuration.md) (in `pyrightconfig.json`/`pyproject.toml`) to point to the install location.

**Comment by vishalovercome** [comment]
I've tried that before creating this issue. Problem is that SynapseML downloads JARs when downloaded via spark shell and I'm not sure how to make it PyLance work with it

**Comment by debonte** [comment]
Looks like `extraPaths` will work if you:
1. Change the extensions of the `.jar` files to `.zip`
2. Add the `.zip` file paths (not the parent directory path) to `extraPaths` (ex. ` "extraPaths": [ "C:/Users/erikd/.ivy2/jars/com.microsoft.azure_synapseml-core_2.12-0.11.2.zip" ]`)

After doing that with `com.microsoft.azure_synapseml-core_2.12-0.11.2.jar`, Pylance was happy with the following import:

```python
from synapse.ml.core.platform.Platform import running_on_synapse
```

We could in theory make Pylance treat `.jar` files like `.zip` and `.egg` files, but this is the first request we've gotten along these lines. Also, I saw the following [in the Spark documentation](https://spark.apache.org/docs/latest/submitting-applications.html#bundling-your-applications-dependencies), which makes me think that SynapseML may be doing something non-standard even for Spark:

> For Python, you can use the `--py-files` argument of spark-submit to add `.py`, `.zip` or `.egg` files to be distributed with your application. If you depend on multiple Python files we recommend packaging them into a `.zip` or `.egg`. For third-party Python dependencies, see [Python Package Management](https://spark.apache.org/docs/latest/api/python/user_guide/python_packaging.html).

**Comment by vishalovercome** [comment]
Having to rename .jar to .zip is a weird work-around. I'll do it now but given that pyspark is such a widely used framework and python itself being often times used as a glue language, I would request you to enhance the extension itself.

**Comment by debonte** [comment]
Ok. I dug into this some more and found that Python's [zip import support](https://peps.python.org/pep-0273/) doesn't care about the extension of zipped files, so ideally we wouldn't either.

Resources:

- [Python docs](https://docs.python.org/3/library/zipimport.html)
- [CPython implementation docs](https://github.com/python/cpython/blob/main/Doc/library/zipimport.rst)
- [CPython implementation](https://github.com/python/cpython/blob/main/Lib/zipimport.py)
- Our issues/PRs related to zip/egg support:
  - https://github.com/microsoft/vscode-python/issues/2532
  - https://github.com/microsoft/python-language-server/issues/196
  - https://github.com/microsoft/pyrx/issues/1340

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.8.31, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023831-16-August-2023-PreRelease)

**Comment by debonte** [comment]
@vishalovercome, you'll now be able to do what I suggested [above](https://github.com/microsoft/pylance-release/issues/4648#issuecomment-1659408330) but without needing to rename the `.jar` files to `.zip`.
